### PR TITLE
Test fundamental properties of `CoinSelection.performSelection`.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue number

ADP-1037

## Summary

This PR tests fundamental properties of the [`CoinSelection.performSelection`](https://github.com/input-output-hk/cardano-wallet/blob/ba3d8d5c8cde7b7aba81e596d4eea425022c49fd/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs#L122) function.

## Details

In the case where [`performSelection`](https://github.com/input-output-hk/cardano-wallet/blob/ba3d8d5c8cde7b7aba81e596d4eea425022c49fd/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs#L122) successfully produces a [`Selection`](https://github.com/input-output-hk/cardano-wallet/blob/ba3d8d5c8cde7b7aba81e596d4eea425022c49fd/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs#L458) `s` for a given set of [`SelectionConstraints`](https://github.com/input-output-hk/cardano-wallet/blob/ba3d8d5c8cde7b7aba81e596d4eea425022c49fd/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs#L343) `cs` and [`SelectionParams`](https://github.com/input-output-hk/cardano-wallet/blob/ba3d8d5c8cde7b7aba81e596d4eea425022c49fd/lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs#L382) `ps`, we verify that the following properties hold:

* `selectionHasSufficientCollateral cs ps s`
* `selectionHasValidSurplus cs ps s`

## Not included in this PR

This PR doesn't include:
- tests that verify the failure cases of `performSelection`
- tests that verify the minimum collateral amount calculation.

These tests will be included in future PRs.

### CODE REVIEW

File: lib/test-utils/src/Test/QuickCheck/Extra.hs [hunk 1]
```diff
@@ -302,6 +314,22 @@ shrinkInterleaved (a, shrinkA) (b, shrinkB) = interleave
     interleave xs [] = xs
     interleave [] ys = ys
 
+--------------------------------------------------------------------------------
+-- Generating and shrinking natural numbers
+--------------------------------------------------------------------------------
+
+chooseNatural :: (Natural, Natural) -> Gen Natural
+chooseNatural (lo, hi) =
+    choose (intCast lo, intCast hi)
```
Comment (jonathanknowles) [inline lib/test-utils/src/Test/QuickCheck/Extra.hs]: Perhaps we could use [`chooseInteger`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:chooseInteger) here, as the QC documentation claims that this is faster.

File: lib/test-utils/src/Test/QuickCheck/Extra.hs [hunk 2]
```diff
@@ -302,6 +314,22 @@ shrinkInterleaved (a, shrinkA) (b, shrinkB) = interleave
...
+    `suchThatMap`
+    intCastMaybe @Integer @Natural
+
+shrinkNatural :: Natural -> [Natural]
+shrinkNatural n
+    = mapMaybe (intCastMaybe @Integer @Natural)
+    $ shrink
```
Comment (jonathanknowles) [inline lib/test-utils/src/Test/QuickCheck/Extra.hs]: Investigate whether `shrinkIntegral` would be faster / more appropriate here.

File: lib/test-utils/src/Test/QuickCheck/Extra.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/test-utils/src/Test/QuickCheck/Extra.hs]: Fixed in 36fe56c3c4adb531c68c9bde4f7b1d5fa1681438.

File: lib/test-utils/src/Test/QuickCheck/Extra.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/test-utils/src/Test/QuickCheck/Extra.hs]: Fixed in 36fe56c3c4adb531c68c9bde4f7b1d5fa1681438.

**Review by sevanspowell — APPROVED** [review]

**Pull request merged**